### ISSUE TITLE
example with AWS OIDC endpoint gives cacheError

### ISSUE THREAD

**Issue by lcmgh** [MinisculeGirraffe/oidc_jwt_validator#2]
```rust
const OIDC_URL: &str = "https://oidc.eks.eu-central-1.amazonaws.com/id/secret";
const TOKEN: &str = "secret";

type TokenClaims = HashMap<String, serde_json::Value>;

#[tokio::test]
async fn test_token_verification() {
    let client = reqwest::ClientBuilder::new()
        .timeout(Duration::from_secs(2))
        .build()
        .unwrap();

    let mut settings = ValidationSettings::new();

    settings.set_issuer(&[OIDC_URL]);
    settings.set_audience(&["account"]);

    let validator = Validator::new(OIDC_URL, client, Strategy::Automatic, settings)
        .await
        .unwrap();

    let validation_result = validator.validate::<TokenClaims>(TOKEN).await;

    println!("{:?}", validation_result);
}
```

gives

`Err(CacheError)`

**Comment by MinisculeGirraffe** [comment]
Let me take a look. I didn't mean this to be like an insanely serious library, even though the code *should* work.

**Comment by MinisculeGirraffe** [comment]
So the reason it doesn't work, at least from what I can see if that directly calling the URL `https://oidc.eks.eu-central-1.amazonaws.com/id/secret`

returns 

`{"message":"Missing Authentication Token"}`

Meaning the library couldn't cache the JWKS because it didn't return a valid  JWKS response. 

The automatic strategy will respect the cache-control & stale while revalidate header present on the response. If the cache is past the expiration time, we pause the incoming requests to validate until it can be refreshed. If it can't be refreshed we return that CacheError. 

So the only situation that will ever occur in is either when the first request fails, in which case it should attempt to re-validate it lazily, one request at a time until the request succeeds.  Or when it previously was successful, and we can't re-validate it after the cache has completely expired.

**Comment by MinisculeGirraffe** [comment]
If there's a specific change I can make to allow supporting authorization parameters on the JWKS endpoint I can. I'm familiar with the AWS/Smithy stuff, i'd just want to confirm how exactly the right auth strategy should go.

If you're just testing stuff out https://openidconnect.net/ is a good tool to just validate behavior.

**Comment by MinisculeGirraffe** [comment]
I sat on the issue for a bit and I think I know a good enough solve for the problem.

Since we accept a reqwest::client, and we're using that to generate an HTTP request. 

We could accept a function that provides a mutable reference to that request before it's signed.  That way if you need to do something to it for authorization you're able to, and the logic doesn't have to change for every kind of implementation. 

You could provide a closure that will sigv4 sign the request, and it should just work.

**Comment by lcmgh** [comment]
Oh your library definitely fills a need. I am trying to implement machine2machine authentication with auth tokens provided by EKS/Kubernetes. Unfortunately I am not sure which auth token is expected in here. I have to check.

**Comment by MinisculeGirraffe** [comment]
In your use case it would need to be sigv4 signed. Typically this involves providing a mutable reference to the standard hyper HTTP request struct, which should be pretty easy.

https://docs.rs/aws-sig-auth/latest/aws_sig_auth/signer/struct.SigV4Signer.html

It may make more sense to incorporate this as a [tower::Service](https://docs.rs/tower/latest/tower/trait.Service.html). I'm still playing a bit with the most ergonomic way of solving the problem.

If you want an easy peasy fix, you can just directly modify the requests [here](https://github.com/MinisculeGirraffe/oidc_jwt_validator/blob/master/src/lib.rs#L79-L119) to include the sigv4 signing step and it would work. It just wouldn't be a generic implementation.

**Comment by lcmgh** [comment]
@MinisculeGirraffe thanks for all your help. Do you know which IAM Role / Policy one needs to sign that request?

**Comment by MinisculeGirraffe** [comment]
I'm not a huge kubernetes expert, so I may be the wrong person to answer this. But I looked a little deeper into it and have concluded I don't know how the heck that this stuff works. 

I'm at least 70% sure you can't even use that URL for that purpose. 

I think this blog post covers the machine2machine auth use case.

https://medium.com/in-the-weeds/service-to-service-authentication-on-kubernetes-94dcb8216cdc

**Comment by MinisculeGirraffe** [comment]
The change may not even be needed.  I looked through the [RFC](https://datatracker.ietf.org/doc/html/rfc8414#section-2) real quick, and the jwks_uri should be a public document served over https. I don't think there should ever be any form of access control on the document since it's just a bunch of public keys in JSON.

More sure now that the URL is just not intended to be used for that purpose.

**Comment by lcmgh** [comment]
Maybe I got a wrong understanding here. I understood that the flow must be like this:

* Retreive JWT token via Kubernetes token injection (this works, I got a valid token)
* Use this library to validate the JWT token at the OIDC endpoint of the token provider

**Comment by lcmgh** [comment]
I started another trial with a different url `const OIDC_URL: &str = "https://oidc.eks.eu-central-1.amazonaws.com/id/<cluster-id>/.well-known/openid-configuration";`

which returns
```json
{"issuer":"https://oidc.eks.eu-central-1.amazonaws.com/id/<cluster-id>","jwks_uri":"https://oidc.eks.eu-central-1.amazonaws.com/id/<cluster-id>/keys","authorization_endpoint":"urn:kubernetes:programmatic_authorization","response_types_supported":["id_token"],"subject_types_supported":["public"],"claims_supported":["sub","iss"],"id_token_signing_alg_values_supported":["RS256"]}
```

The code then fails with

```
thread 'api::k8sauth::test_token_verification' panicked at 'called `Result::unwrap()` on an `Err` value: RequestFailed(reqwest::Error { kind: Decode, source: Error("missing field `jwks_uri`", line: 1, column: 42) })', 
```

Don't really understand why it fails because the field `jwks_uri` is included in the response.

**Comment by MinisculeGirraffe** [comment]
Try using 

`https://oidc.eks.eu-central-1.amazonaws.com/id/<cluster-id>`

and it should work then. 

I'm doing some very lazy detection and parsing of the url, so it's expecting a very specific format. I can address this and make it a little nicer to use.

**Comment by MinisculeGirraffe** [comment]
Published a new version, 0.1.1, that will accept either 
`https://oidc.eks.eu-central-1.amazonaws.com/id/<cluster-id>`
or 
`https://oidc.eks.eu-central-1.amazonaws.com/id/<cluster-id>/.well-known/openid-configuration`

**Comment by lcmgh** [comment]
Hi @MinisculeGirraffe . Thanks for the update. Unfortunately I am still facing: Validation result: Err(CacheError)

Could you elaborate on that error?

**Comment by lcmgh** [comment]
I found an example OIDC endpoint on Github: https://github.com/epmd-edp/edp-terraform-aws-platform/blob/c1ffec00edf4615ec60c9bf369fb3489ce3865b3/modules/iam/examples/kaniko/variables.tf#L20

You can reproduce the issue with that url.

**Comment by MinisculeGirraffe** [comment]
Same deal, it's not returning a valid OIDC discovery response.

URL Provided to the library: `https://oidc.eks.eu-central-1.amazonaws.com/id/01987654321`
Discovery URL library calls: `https://oidc.eks.eu-central-1.amazonaws.com/id/01987654321/.well-known/openid-configuration`

Response library gets from the discovery url.
```
{"id":"01987654321","message":"ID is not found"}
```

Now, That is a bad error message and isn't exactly clear. I can add another value to the enum to cover the condition of when the discovery fails. Since it's not really a Cache error at that point. 

Edit: looks like I already have one, i'm just not using it correctly. I'll write a fix for this.

**Comment by MinisculeGirraffe** [comment]
I'll try and spin up an EKS cluster later today to see if I can replicate it in an environment. By all counts, if a valid openid discovery document, and jwks is returned, the library should work regardless of environment.

**Comment by lcmgh** [comment]
Sorry for providing a link to a non-existent EKS cluster. 

I just checked the response to the well known URL in Chrome. Might it because there is no cache header in the response?

```
Content-Length: 419
Content-Type: application/json
Date: Wed, 05 Jul 2023 16:04:53 GMT
X-Amz-Apigw-Id: HmQp2F*****
X-Amzn-Requestid: bfd0ed7a-46ca-4161-9db8-*****
X-Amzn-Trace-Id: Root=1-64a594a5-*****
```

**Comment by lcmgh** [comment]
So I gave it a try with manual Cache settings

```rust
    let validator = Validator::new(
        OIDC_URL,
        client,
        Strategy::Manual(Settings {
            max_age: Duration::from_secs(60),
            stale_while_revalidate: None,
            stale_if_error: None,
        }),
        settings,
    )
    .await
    .unwrap();
```

But error remains `Validation result: Err(CacheError)`

**Comment by MinisculeGirraffe** [comment]
In the case that there's no cache-control headers present, it defaults to a 1 second lazy polling interval on the JWKS.  That tends to be the default behavior most libraries from other languages do in that scenario, so that's what I implemented. Might want to document that better lmao.

A manual strategy isn't required, but it's available if you'd like to poll less often than 1 second when the cache-control headers aren't present. 

Typically though respecting the cache-control headers is the best method. But you don't always control what the OIDC provider does.

**Comment by MinisculeGirraffe** [comment]
Huh. Just replicated it with my own EKS cluster. That's wild. One sec let me figure this out

**Comment by MinisculeGirraffe** [comment]
Oh fuck that's a dumb bug.

**Comment by MinisculeGirraffe** [comment]
Alright i've pushed a fix and published v0.2 which contains the changes +  some doc updates.

The TL:DR of why that was happening. 

It was correct to return an error, it was just the wrong error kind. 

The JWKS contains the keys used to validate a JWT. They're indexed in a hashmap by the KID field, so they can be looked up quickly.

When we get an incoming token for validation, we check if it has the kid field present, if it doesn't we early return an error, because there's no point on checking the signature on that token.

Once we know the token has a KID, we check it against our cached values to see if there's a JWK with that matching key id.  
If there isn't a matching JWK, we trigger a lazy re-validation to make sure that an authentication isn't erroneously denied because the JWKS behind the cache had changed. Only 1 network request will ever be sent at a time when this happens as to not have it be a DOS vector.

The function signature for this originally being
```rust
   async fn get_kid_retry(&self, kid: impl AsRef<str>) -> Result<Arc<DecodingInfo>, ValidationError> 
```

However it actually needed to be 

```rust
   async fn get_kid_retry(&self, kid: impl AsRef<str>) -> Result<Option<Arc<DecodingInfo>>, ValidationError> 
```

Since the key we were looking for is allowed to still not be present, even after a cache re-validation. I was erroneously returning a cache error when it wasn't present after revalidation.

It still should be an error, just a different one.  So I added another error kind for that. 



Commit for the details, it's a small change

https://github.com/MinisculeGirraffe/oidc_jwt_validator/commit/aba3af524455f4a966402cb22bf7f46a44ffbf0d

**Comment by lcmgh** [comment]
I can confirm it works :) Thanks so much! I will reach out to AWS regarding the caching header :D

**Comment by lcmgh** [comment]
@MinisculeGirraffe I stand corrected that AWS does not send a cache header. This was the case on the well known config page but `https://oidc.eks.eu-central-1.amazonaws.com/id/<cluster-id>/keys` comes indeed with a `Cache-Control` header.

**Comment by lcmgh** [comment]
I thought about a HashMap workaround when using the "normalized" OIDC Provider URL as Key:

validator: HashMap<String, Validator>
But when trying to retrieve the verifier my code is not aware of the issuer url as the token was not yet decoded.
